core/core/test/serverspec/spec/zookeeper/zookeeper_spec.rb

```diff
@@ -1,27 +1,86 @@
 require 'spec_helper'
 
-describe service('zookeeper') do
-  it { should be_enabled }
-  it { should be_running }
+describe 'Checking if zookeeper service is up' do
+  describe service('zookeeper') do
+    it { should be_enabled }
+    it { should be_running }
+  end
 end
 
-describe port(2181) do
-  it { should be_listening }
+describe 'Checking if the ports are open' do
+  describe port(2181) do            #2181 for client connections
+    let(:disable_sudo) { false }  
+    it { should be_listening }
+  end
+
+  # These ports are not always open. Depends on configuration.
+  # describe port(2888) do          #2888 for follower connections
+  #   let(:disable_sudo) { false }
+  #   it { should be_listening }
+  # end
+  # describe port(3888) do          #3888 for inter nodes connections
+  #   let(:disable_sudo) { false }
+  #   it { should be_listening }
+  # end
+
+  #checking port 2888
+  describe command('if /opt/zookeeper/bin/zkServer.sh status | grep "Mode: leader"; then netstat -tunl | grep 2888; else echo "not leader"; fi') do
+    its(:stdout) { should match /2888|not leader/ }
+    its(:exit_status) { should eq 0 }
+  end
+
+  #checking port 3888
+  describe command('if /opt/zookeeper/bin/zkServer.sh status | grep "Mode: standalone"; then echo "standalone"; else netstat -tunl | grep 3888; fi') do
+    its(:stdout) { should match /3888|standalone/ }
+    its(:exit_status) { should eq 0 }
+  end
+end  
+
+
+describe 'Checking user configuration' do
+  describe group('zookeeper') do
+    it { should exist }
+  end
+  describe user('zookeeper') do
+    it { should exist }
+    it { should belong_to_group 'zookeeper' }
+    it { should have_home_directory '/home/zookeeper' }
+    it { should have_login_shell '/sbin/nologin' }
```

Comment:
Should be: /usr/sbin/nologin
It'll work on RHEL7 because its /sbin is linked to /usr/sbin. On Ubuntu it's not linked (at least Azure images of 18.04).

Either way /usr/sbin should be safe for distros made in the last 5 years or so.

Reply:
Please note this change: https://github.com/epiphany-platform/epiphany/pull/97